R/derive_joined.R

```diff
@@ -0,0 +1,334 @@
+#' ) %>%
+#'   rename(APERSDT = S, APEREDT = E) %>%
+#'   mutate(
+#'     APERIOD = as.integer(APERIODC),
+#'     APERIODC = paste("Period", APERIODC)
+#'   )
+#'
+#' period_ref
+#'
+#' adae <- tribble(
+#'   ~USUBJID, ~ASTDT,
+#'   "1",      "2021-01-01",
+#'   "1",      "2021-01-05",
+#'   "1",      "2021-02-05",
+#'   "1",      "2021-03-05",
+#'   "1",      "2021-04-05",
+#'   "2",      "2021-02-15",
+#' ) %>%
+#'   mutate(
+#'     ASTDT = ymd(ASTDT)
+#'   )
+#'
+#' derive_vars_joined(
+#'   adae,
+#'   dataset_add = period_ref,
+#'   by_vars = vars(USUBJID),
+#'   new_vars = vars(APERIOD, APERIODC),
+#'   join_vars = vars(APERSDT, APEREDT),
+#'   filter_join = APERSDT <= ASTDT & ASTDT <= APEREDT
+#' )
+derive_vars_joined <- function(dataset,
+                               dataset_add,
+                               by_vars = NULL,
+                               order = NULL,
+                               new_vars = NULL,
+                               join_vars = NULL,
+                               filter_add = NULL,
+                               filter_join = NULL,
+                               mode = NULL,
+                               check_type = "warning") {
+  assert_vars(by_vars, optional = TRUE)
+  by_vars_left <- replace_values_by_names(by_vars)
+  assert_order_vars(order, optional = TRUE)
+  assert_vars(new_vars, optional = TRUE)
+  assert_vars(join_vars, optional = TRUE)
+  assert_data_frame(dataset, required_vars = by_vars_left)
+  assert_data_frame(
+    dataset_add,
+    required_vars = quo_c(by_vars, join_vars, extract_vars(order), new_vars)
+  )
+  filter_add <- assert_filter_cond(enquo(filter_add), optional = TRUE)
+  filter_join <- assert_filter_cond(enquo(filter_join), optional = TRUE)
+
+  if (is.null(new_vars)) {
+    new_vars <- chr2vars(colnames(dataset_add))
+  }
+
+  # number observations of the input dataset to get a unique key
+  # (by_vars and tmp_obs_nr)
+  tmp_obs_nr <- get_new_tmp_var(dataset, prefix = "tmp_obs_nr")
+  data <- dataset %>%
+    derive_var_obs_number(
+      new_var = !!tmp_obs_nr,
+      by_vars = by_vars_left,
+      check_type = "none"
+    )
+
+  # prepare right side of the join,
+  # by_vars are renamed here, new_vars will be renamed at the end
+  data_right <- filter_if(dataset_add, filter_add) %>%
+    select(!!!by_vars, !!!join_vars, !!!unname(new_vars))
+
+  # join dataset (if no by variable, a full join is performed)
+  data_joined <- left_join(
+    data,
+    data_right,
+    by = vars2chr(by_vars_left),
+    suffix = c("", ".join")
+  )
+
+  # select observations for the new variables
+  data_return <- filter_if(data_joined, filter_join)
+
+  common_vars <-
+    chr2vars(setdiff(intersect(colnames(data), colnames(data_right)), vars2chr(by_vars)))
+  if (!is.null(order)) {
+    data_return <- filter_extreme(
+      data_return,
+      by_vars = quo_c(by_vars_left, quo(!!tmp_obs_nr)),
+      order = add_suffix_to_vars(order, vars = common_vars, suffix = ".join"),
```

Comment:
When `dataset` and `dataset_add` has the same variable names (e.g. when joining a dataset to itself), should we explain which dataset the `order` parameter is referring to? It's clear in the codes, but should we add in the documentation the first/last obs are chosen based on the variable joined from `dataset_add` after applying `filter_join`? If you think it's very self-explanatory and adding an explanation causes more confusion I'm happy to leave as it is.

Reply:
You could deduce it from the description but it is definitely not obvious. Thus it is a good idea to add it.